src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs

```diff
@@ -24,6 +24,61 @@
+    ///     [KeyType(Contiguous = true, Count = m, Min = firstRowIndex)]
+    ///     public uint MatrixRowIndex;
+    ///     // The rating at the MatrixColumnIndex-th column and the MatrixRowIndex-th row.
+    ///     public float Value;
+    /// }
+    /// </code>
+    /// Notice that it's not necessary to specify all entries in the training matrix, so matrix factorization can be used to fill <i>missing values</i>.
+    /// This behavior is very helpful when building recommender systems.</p>
+    /// <p>To provide a better understanding on practical uses of matrix factorization, let's consider music recommendation as an example.
+    /// Assume that user IDs and music IDs are used as row and column indexes, respectively, and matrix's values are ratings provided by those users. That is,
+    /// rating <i>r</i> at row <i>r</i> and column <i>v</i> means that user <i>u</i> give <i>r</i> to item <i>v</i>.
+    /// An imcomplete matrix is very common because not all users may provide their feedbacks to all products (for example, no one can rate ten million songs).
+    /// Assume that<i>R</i> is a m-by-n rating matrix and the rank of the two factor matrices are<i>P</i> (m-by-k matrix) and <i>Q</i> (n-by-k matrix), where k is the approximation rank.
+    /// The predicted rating at the u-th row and the v-th column in <i>R</i> would be the inner product of the u-th row of P and the v-th row of Q; that is,
+    /// <i>R</i> is approximated by the product of <i>P</i>'s transpose and <i>Q</i>. This trainer implements
+    /// <a href='https://www.csie.ntu.edu.tw/~cjlin/papers/libmf/mf_adaptive_pakdd.pdf'>a stochastic gradient method</a> for finding <i>P</i>
+    /// and <i>Q</i> via minimizing the distance between<i> R</i> and the product of <i>P</i>'s transpose and Q.</p>.
+    /// <p>For users interested in the mathematical details, please see the references below.
+    ///     <list type = 'bullet'>
+    ///         <item>
+    ///             <description><a href='https://www.csie.ntu.edu.tw/~cjlin/papers/libmf/libmf_journal.pdf' > A Fast Parallel Stochastic Gradient Method for Matrix Factorization in Shared Memory Systems</a></description>
+    ///         </item>
+    ///         <item>
+    ///             <description><a href='https://www.csie.ntu.edu.tw/~cjlin/papers/libmf/mf_adaptive_pakdd.pdf' > A Learning-rate Schedule for Stochastic Gradient Methods to Matrix Factorization</a></description>
+    ///         </item>
+    ///         <item>
+    ///             <description><a href='https://www.csie.ntu.edu.tw/~cjlin/papers/libmf/libmf_open_source.pdf' > LIBMF: A Library for Parallel Matrix Factorization in Shared-memory Systems</a></description>
+    ///         </item>
+    ///     </list>
+    /// </p>
+    /// <p>Example code can be found by searching for <i>MatrixFactorization</i> in <a href='https://github.com/dotnet/machinelearning'>ML.NET.</a></p>
```

Comment:
><p>Example code can be found by searching for <i>MatrixFactorization</i> in <a href='https://github.com/dotnet/machinelearning'>ML.NET.</a></p> [](start = 8, length = 143)

add the reference, so it displays inline?

Reply:
Do you know how to do that?

---
In reply to: [229499010](https://github.com/dotnet/machinelearning/pull/1450#discussion_r229499010) [](ancestors = 229499010)

Reply:
similar to here:

[https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.StandardLearners/Standard/SdcaStatic.cs#L41](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.StandardLearners/Standard/SdcaStatic.cs#L41)



---
In reply to: [229504220](https://github.com/dotnet/machinelearning/pull/1450#discussion_r229504220) [](ancestors = 229504220,229499010)

Reply:
Thanks!

---
In reply to: [229507236](https://github.com/dotnet/machinelearning/pull/1450#discussion_r229507236) [](ancestors = 229507236,229504220,229499010)